### ISSUE TITLE
0xDetermination - A borrower eligible for liquidation can pay an improperly large amount of fees, and may be unfairly liquidated

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-02-leverage-contracts-judging#40]
0xDetermination

high

# A borrower eligible for liquidation can pay an improperly large amount of fees, and may be unfairly liquidated

## Summary
If a borrower is partially liquidated and then increases the collateral balance to avoid further liquidation, they will pay an improperly large amount of fees and can be unfairly liquidated.
## Vulnerability Detail
The root cause is that partial emergency liquidation doesn't update `accLoanRatePerSeconds` (https://github.com/sherlock-audit/2024-02-leverage-contracts/blob/main/wagmi-leverage/contracts/LiquidityBorrowingManager.sol#L660-L666).

If a borrower is partially liquidated, fees will be increased by the entire collateral amount (https://github.com/sherlock-audit/2024-02-leverage-contracts/blob/main/wagmi-leverage/contracts/LiquidityBorrowingManager.sol#L604-L633):
```solidity
            (collateralBalance, currentFees) = _calculateCollateralBalance(
                borrowing.borrowedAmount,
                borrowing.accLoanRatePerSeconds,
                borrowing.dailyRateCollateralBalance,
                accLoanRatePerSeconds
            );
            ...
            if (collateralBalance > 0) {
                ...
            } else {
                currentFees = borrowing.dailyRateCollateralBalance; //entire collateral amount
            }
            ...
            borrowing.feesOwed += _pickUpPlatformFees(borrowing.holdToken, currentFees);
```
When liquidation occurs right after becoming liquidatable, the `collateralBalance` calculation in `repay()` above will be a small value like `-1`; and essentially all the fees owed will be collected.

If the borrower notices the partial liquidation and wishes to avoid further liquidation, `increaseCollateralBalance()` can be called to become solvent again. But since the `accLoanRatePerSeconds` wasn't updated, the borrower will have to doubly pay all the fees that were just collected. This will happen if a lender calls `harvest()` or the loan is liquidated again. The loan can also be liquidated unfairly, because the `collateralBalance` calculated above will be much lower than it should be.
## Impact
The borrower may pay too many fees, and it's also possible to unfairly liquidate the position.
## Code Snippet
https://github.com/sherlock-audit/2024-02-leverage-contracts/blob/main/wagmi-leverage/contracts/LiquidityBorrowingManager.sol#L660-L666
https://github.com/sherlock-audit/2024-02-leverage-contracts/blob/main/wagmi-leverage/contracts/LiquidityBorrowingManager.sol#L604-L633
## Tool used

Manual Review

## Recommendation
Update `accLoanRatePerSeconds` for incomplete emergency liquidations.

**Comment by fann95** [comment]
accLoanRatePerSeconds should not be updated since borrowedAmount are reduced, accordingly, a position debt also redused
This issue was already discussed in the previous audit.
Harvest cannot be called on a position that is under liquidation

**Comment by 0xDetermination** [comment]
@nevillehuang @Czar102 The root cause and fix for this issue are both distinct from #41- the erroneous distribution of fees in #41 is caused by not harvesting fees when a new loan is taken with the `borrow()` function, and the fix is the new internal `_harvest()` function that runs in `borrow()`. ([link](https://github.com/RealWagmi/wagmi-leverage/commit/84416fcedfcc7eb062917bdc69f919bba9d3c0b7#diff-19eb89909078be53347656861a11ad5cc119624558cf955fc423c728d80821bdR976))

This issue #40 is caused by the fee distribution mechanism in emergency liquidation mode in `repay()`, which is separate from `borrow()` and `harvest()`. The root cause and fix can both be seen [here](https://github.com/RealWagmi/wagmi-leverage/commit/84416fcedfcc7eb062917bdc69f919bba9d3c0b7#diff-19eb89909078be53347656861a11ad5cc119624558cf955fc423c728d80821bdL663-R604) in `repay()`. The fix for #41 won't fix this issue.

**Comment by 0xDetermination** [comment]
Escalate

I understand @nevillehuang's point here, but I still think this shouldn't be a dup for the reasons I gave in my above [comment](https://github.com/sherlock-audit/2024-02-leverage-contracts-judging/issues/40#issuecomment-1976611101). The fix PR changed a lot of things unrelated to validated issues, such as removal of the min fee [mechanism](https://github.com/RealWagmi/wagmi-leverage/commit/84416fcedfcc7eb062917bdc69f919bba9d3c0b7#diff-19eb89909078be53347656861a11ad5cc119624558cf955fc423c728d80821bdL618-L624).

Will appreciate @Czar102's decision.

Additionally, if validated as not a dup, not sure if this should be H or M based on the 'external conditions' criteria.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> I understand @nevillehuang's point here, but I still think this shouldn't be a dup for the reasons I gave in my above [comment](https://github.com/sherlock-audit/2024-02-leverage-contracts-judging/issues/40#issuecomment-1976611101). The fix PR changed a lot of things unrelated to validated issues, such as removal of the min fee [mechanism](https://github.com/RealWagmi/wagmi-leverage/commit/84416fcedfcc7eb062917bdc69f919bba9d3c0b7#diff-19eb89909078be53347656861a11ad5cc119624558cf955fc423c728d80821bdL618-L624).
> 
> Will appreciate @Czar102's decision.
> 
> Additionally, if validated as not a dup, not sure if this should be H or M based on the 'external conditions' criteria.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by Czar102** [comment]
@nevillehuang can you elaborate on how does a fix to #40 follows from #41?

**Comment by nevillehuang** [comment]
@Czar102 I believe @fann95 answered it [here](https://github.com/sherlock-audit/2024-02-leverage-contracts-judging/issues/40#issuecomment-1976134206) and [here](https://github.com/sherlock-audit/2024-02-leverage-contracts-judging/issues/40#issuecomment-1974030831)

Based on sponsor description and comments above, The PR fix was performed without consideration of this issue, which led me to believe this issue only stems from the root cause of incorrect distribution.

**Comment by Czar102** [comment]
@nevillehuang what is the single logical error (maybe an assumption, maybe approach) that led to both of these issues?

Having issue A which makes the sponsor want to restructure code and it accidentally removing issue B doesn't make them duplicates.

**Comment by fann95** [comment]
I propose to confirm this issue and not consider it a duplicate.

**Comment by nevillehuang** [comment]
@Czar102 It stems from the logic for fee distribution. Although I disagree, seems like sponsor agrees to deduplicate so we can proceed with deduplication.

**Comment by Czar102** [comment]
@nevillehuang regarding this:
> It stems from the logic for fee distribution.

I believe it doesn't answer my question:
> what is the single logical error (maybe an assumption, maybe approach) that led to both of these issues?

I wanted to have a one-sentence description of the common ground of these issues, and the fact that the issues "stem from the logic for fee distribution" (are in the same part of the code logic) doesn't make them duplicates.

I'm planning to make this issue a unique issue, unless a justification (as mentioned above) is provided.

What are the considerations regarding the severity? @nevillehuang
@0xDetermination what did you mean by the following fragment of your escalation?
> not sure if this should be H or M based on the 'external conditions' criteria

**Comment by 0xDetermination** [comment]
@Czar102 I'm not 100% sure whether this is better suited for H or M, as I don't have a ton of experience with Sherlock judging rules. Basically, the issue can cause serious loss of funds (borrower's entire collateral), but it is conditional on a partial emergency liquidation followed by the borrower increasing collateral. It looks more like M to me but I don't want to speak too soon, will leave it up to you and @nevillehuang. Happy to provide more info if needed.

**Comment by nevillehuang** [comment]
Yup @Czar102 I agree with your decision, I cannot pinpoint an exact singular approach/code logic given this is an update contest and would take up too much time. The reason I duplicated them was on the side of caution, given the sponsor quite literally fix this issue without even considering it. But since sponsor also agree with deduplication, lets move ahead

**Comment by Czar102** [comment]
In that case, planning to make this issue a unique Medium severity one.

**Comment by Czar102** [comment]
Result:
Medium
Unique

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [0xDetermination](https://github.com/sherlock-audit/2024-02-leverage-contracts-judging/issues/40/#issuecomment-1986871881): accepted

**Comment by 0xDetermination** [comment]
Fix looks good, collection of fees in emergency mode `repay()` has been reworked. Notably, `dailyRateCollateralBalance` is no longer set to zero.

**Comment by sherlock-admin4** [comment]
The Lead Senior Watson signed off on the fix.
